File: SEALNETTest/ContextWrapper.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System;
using Microsoft.Research.SEAL;

namespace SEALNETTest
{
    [TestClass]
    public class ContextWrapper
    {
        [TestMethod]
        public void ContextConstructorNET()
        {
            // Nothing set
            var parms = new EncryptionParameters();
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsFalse(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Not relatively prime coeff moduli
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = new List<SmallModulus> { 2, 30 };
            parms.PlainModulus = 2;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsFalse(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Plain modulus not relatively prime to coeff moduli
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = new List<SmallModulus> { 17, 41 };
            parms.PlainModulus = 34;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsFalse(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Plain modulus not smaller than product of coeff moduli
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 2 });
            parms.PlainModulus = 3;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus[0].ToString() == "2");
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsFalse(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // FFT poly but not NTT modulus
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 3 });
            parms.PlainModulus = 2;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus[0].ToString() == "3");
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Parameters OK; no fast plain lift
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 17, 41 });
            parms.PlainModulus = 18;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus.ToString().Equals("2B9"));
                Assert.IsTrue(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsTrue(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Parameters OK; fast plain lift
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 17, 41 });
            parms.PlainModulus = 16;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus.ToString().Equals("2B9"));
                Assert.IsTrue(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsTrue(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsTrue(qualifiers.EnableFastPlainLift);
            }

            // Parameters OK; no batching due to non-prime plain modulus
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 17, 41 });
            parms.PlainModulus = 49;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus.ToString().Equals("2B9"));
                Assert.IsTrue(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsTrue(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Parameters OK; batching enabled
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 17, 41 });
            parms.PlainModulus = 73;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus.ToString().Equals("2B9"));
                Assert.IsTrue(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsTrue(qualifiers.EnableNTT);
                Assert.IsTrue(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }

            // Parameters OK; batching and fast plain lift enabled
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 137, 193 });
            parms.PlainModulus = 73;
            parms.NoiseStandardDeviation = 3.19;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsTrue(context.TotalCoeffModulus.ToString().Equals("6749"));
                Assert.IsTrue(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsTrue(qualifiers.EnableNTT);
                Assert.IsTrue(qualifiers.EnableBatching);
                Assert.IsTrue(qualifiers.EnableFastPlainLift);
            }

            // Negative noise standard deviation
            parms.PolyModulus = "1x^4 + 1";
            parms.CoeffModulus = (new List<SmallModulus> { 137, 193 });
            parms.PlainModulus = 73;
            parms.NoiseStandardDeviation = -0.1;
            {
                var context = new SEALContext(parms);
                var qualifiers = context.Qualifiers;
                Assert.IsFalse(qualifiers.ParametersSet);
                Assert.IsTrue(qualifiers.EnableFFT);
                Assert.IsFalse(qualifiers.EnableNTT);
                Assert.IsFalse(qualifiers.EnableBatching);
                Assert.IsFalse(qualifiers.EnableFastPlainLift);
            }
        }
    };
}

File: SEALNET/sealnet/ContextWrapper.cpp
#include "sealnet/ContextWrapper.h"
#include "sealnet/EncryptionParamsWrapper.h"
#include "sealnet/Common.h"
#include "sealnet/SmallModulusWrapper.h"

using namespace System;
using namespace System::IO;
using namespace std;
using namespace System::Collections::Generic;

namespace Microsoft
{
    namespace Research
    {
        namespace SEAL
        {
            EncryptionParameterQualifiers::EncryptionParameterQualifiers(const seal::EncryptionParameterQualifiers &qualifiers) : 
                qualifiers_(nullptr)
            {
                try
                {
                    qualifiers_ = new seal::EncryptionParameterQualifiers(qualifiers);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            EncryptionParameterQualifiers::~EncryptionParameterQualifiers()
            {
                this->!EncryptionParameterQualifiers();
            }

            EncryptionParameterQualifiers::!EncryptionParameterQualifiers()
            {
                if (qualifiers_ != nullptr)
                {
                    delete qualifiers_;
                    qualifiers_ = nullptr;
                }
            }

            seal::EncryptionParameterQualifiers &EncryptionParameterQualifiers::GetQualifiers()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return *qualifiers_;
            }

            bool EncryptionParameterQualifiers::ParametersSet::get()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return qualifiers_->parameters_set;
            }

            bool EncryptionParameterQualifiers::EnableFFT::get()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return qualifiers_->enable_fft;
            }

            bool EncryptionParameterQualifiers::EnableNTT::get()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return qualifiers_->enable_ntt;
            }

            bool EncryptionParameterQualifiers::EnableBatching::get()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return qualifiers_->enable_batching;
            }

            bool EncryptionParameterQualifiers::EnableFastPlainLift::get()
            {
                if (qualifiers_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("EncryptionParameterQualifiers is disposed");
                }
                return qualifiers_->enable_fast_plain_lift;
            }

            SEALContext::SEALContext(EncryptionParameters ^parms)
            {
                if (parms == nullptr)
                {
                    throw gcnew ArgumentNullException("parms cannot be null");
                }
                try
                {
                    context_ = new seal::SEALContext(parms->GetParms());
                    GC::KeepAlive(parms);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            SEALContext::SEALContext(EncryptionParameters ^parms, MemoryPoolHandle ^pool)
            {
                if (parms == nullptr)
                {
                    throw gcnew ArgumentNullException("parms cannot be null");
                }
                if (pool == nullptr)
                {
                    throw gcnew ArgumentNullException("pool cannot be null");
                }
                try
                {
                    context_ = new seal::SEALContext(parms->GetParms(), pool->GetHandle());
                    GC::KeepAlive(parms);
                    GC::KeepAlive(pool);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            SEALContext::SEALContext(SEALContext ^copy)
            {
                if (copy == nullptr)
                {
                    throw gcnew ArgumentNullException("copy cannot be null");
                }
                try
                {
                    context_ = new seal::SEALContext(copy->GetContext());
                    GC::KeepAlive(copy);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            SEALContext::SEALContext(const seal::SEALContext &context) : context_(nullptr)
            {
                try
                {
                    context_ = new seal::SEALContext(context);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            EncryptionParameters ^SEALContext::Parms::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return gcnew EncryptionParameters(context_->parms());
            }

            EncryptionParameterQualifiers ^SEALContext::Qualifiers::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return gcnew EncryptionParameterQualifiers(context_->qualifiers());
            }

            BigPoly ^SEALContext::PolyModulus::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return gcnew BigPoly(context_->poly_modulus());
            }

            List<SmallModulus^> ^SEALContext::CoeffModulus::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                auto result = gcnew List<SmallModulus^>;
                for (auto mod : context_->coeff_modulus())
                {
                    result->Add(gcnew SmallModulus(mod));
                }
                return result;
            }

            SmallModulus ^SEALContext::PlainModulus::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return gcnew SmallModulus(context_->plain_modulus());
            }

            double SEALContext::NoiseStandardDeviation::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return context_->noise_standard_deviation();
            }

            double SEALContext::NoiseMaxDeviation::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return context_->noise_max_deviation();
            }

            BigUInt ^SEALContext::TotalCoeffModulus::get()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return gcnew BigUInt(context_->total_coeff_modulus());
            }

            seal::SEALContext &SEALContext::GetContext()
            {
                if (context_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("SEALContext is disposed");
                }
                return *context_;
            }

            SEALContext::~SEALContext()
            {
                this->!SEALContext();
            }

            SEALContext::!SEALContext()
            {
                if (context_ != nullptr)
                {
                    delete context_;
                    context_ = nullptr;
                }
            }
        }
    }
}